### PR TITLE
Rkuris/range proofs

### PR DESCRIPTION
**Pull request by rkuris**
The generic N can't be specified by the caller, since the method needs to create owned things of type N, so this generic was removed from the trait.

There is a subtle difference between `Into::into` and `Box::new` in that the former is not Send for some reason.

### CODE REVIEW

File: firewood/src/db.rs
Comment (richardpringle) [inline firewood/src/db.rs:323]:
```suggestion
        let Some((key, _)) = it.next().await.transpose()? else {
            return Ok(None);
        };
        
        let Some(first_key) = self.single_key_proof(key).await? else {
            return Ok(None);
        };
```

File: firewood/src/db.rs
Comment (richardpringle) [inline firewood/src/db.rs:326]: This comment isn't necessary

File: firewood/src/db.rs
Comment (richardpringle) [inline firewood/src/db.rs:349]: I think what you really want is:
```rust
        let mut middle = stream
            .take_while(|kv_result| {
                let Some(last_key) = last_key.as_ref() else {
                    return ready(true);
                };

                let result = kv_result
                    .as_ref()
                    .map(|(key, _)| key.as_slice() < last_key.as_ref())
                    .unwrap_or(true); // always return Err(_)

                ready(result)
            })
            .take(limit)
            .try_collect::<Vec<_>>()
            .await?;
```

File: firewood/src/db.rs
Comment (richardpringle) [inline firewood/src/db.rs:353]: If there's only ever one message in a `RangeProofError`, we should just have this message as part of the display. 

Alternatively, it's easy to put `&'static str` as the data-type in the Error-variant.

**Review by richardpringle — CHANGES_REQUESTED** [review]

File: firewood/src/db.rs
Comment (rkuris) [reply firewood/src/db.rs:323]: Very nice! Love it!

File: firewood/src/db.rs
Comment (rkuris) [reply firewood/src/db.rs]: This was fixed. This PR isn't really ready for review yet.

File: firewood/src/db.rs
Comment (rkuris) [reply firewood/src/db.rs:349]: I refactored this. Take should probably be first, since it's way cheaper.

File: firewood/src/db.rs
Comment (rkuris) [reply firewood/src/db.rs]: I was able to shorten it quite a bit so I don't know that this is still necessary.

**Review by xinifinity — COMMENTED** [review]
Sorry, was trying to review, but not sure if the PR has the correct base and which code is committed, do you mind doing another rebase with the current code? Thanks!

**Comment by rkuris** [comment]
> Sorry, was trying to review, but not sure if the PR has the correct base and which code is committed, do you mind doing another rebase with the current code? Thanks!

Done! Was a bit of a challenge to rebase so I just cherry picked and force pushed.

File: firewood/src/db.rs
Comment (richardpringle) [inline firewood/src/db.rs:334]: This needs to come after the `take_while`

File: firewood/src/db.rs
Comment (richardpringle) [inline firewood/src/db.rs:309]: This should really be implemented in `merkle.rs` and tested there as well. You should be able to generate a `RangeProof` on `Merkle` and the `Db::range_proof` should just delegate.

File: firewood/src/db.rs
Comment (richardpringle) [reply firewood/src/db.rs:334]: See my comment above about `merkle.rs`. If we put this functionality there, it's also easy to test. we should have tests before merging. Not necessarily tests for proving a range-proof, but tests that give us confidence that the filters work properly (first-key, last-key, and limit).

File: firewood/src/db.rs
Comment (richardpringle) [reply firewood/src/db.rs:349]: Much better

**Review by richardpringle — CHANGES_REQUESTED** [review]
Sorry, it's definitely extra work, but I didn't think about moving this until I thought about tests. And the `take` call is in the wrong place.

File: firewood/src/db.rs
Comment (richardpringle) [reply firewood/src/db.rs:334]: Ah, you're right, my bad. I was thinking of `filter` (my brain was playing tricks on me).

File: firewood/src/db.rs
Comment (xinifinity) [reply firewood/src/db.rs:309]: +1

File: firewood/src/db.rs
Comment (xinifinity) [inline firewood/src/db.rs:358]: nit: last_key_proof.

File: firewood/src/db.rs
Comment (rkuris) [reply firewood/src/db.rs:311]: This is a good point. We fetch the first_key even if it's less than the last_key, and we should just return an empty. I created an issue to track this.

**Comment by rkuris** [comment]
The linter error here is fixed in #355

**Review by xinifinity — APPROVED** [review]

File: firewood/src/db.rs
Comment (richardpringle) [reply firewood/src/db.rs:311]: I think the current behaviour is fine. We basically don't do any work and leave it to the caller to pass in sane values. That's the nice thing about iterators

**Review by richardpringle — APPROVED** [review]

**Pull request merged**
